geps/gep-1762.md

```diff
@@ -0,0 +1,299 @@
+        type: Utilization
+        averageUtilization: 50
+---
+apiVersion: policy/v1
+kind: PodDisruptionBudget
+metadata:
+  name: gateway
+spec:
+  minAvailable: 1
+  selector:
+    # Match the generated Deployment by label
+    matchLabels:
+      gateway.networking.k8s.io/metadata.name: gateway
+```
+
+Note: there is [discussion](https://github.com/kubernetes-sigs/gateway-api/discussions/1355) around a way to attach a HPA to a Gateway directly.
+
+## API
+
+This GEP extends the `infrastructure` API introduced in [GEP-1867](https://gateway-api.sigs.k8s.io/geps/gep-1867).
+
+```go
+type GatewayInfrastructure struct {
+	// Labels that should be applied to any resources created in response to this Gateway.
+	//
+	// For implementations creating other Kubernetes objects, this should be the `metadata.labels` field on resources.
+	// For other implementations, this refers to any relevant (implementation specific) "labels" concepts.
+	//
+	// If Labels is set on the GatewayClass as well, the labels are merged with the Gateway's labels taking precedence.
+	//
+	// Support: Extended
+	Labels map[string]string `json:"labels,omitempty"`
+	// Annotations that should be applied to any resources created in response to this Gateway.
+	//
+	// For implementations creating other Kubernetes objects, this should be the `metadata.annotations` field on resources.
+	// For other implementations, this refers to any relevant (implementation specific) "annotations" concepts.
+	//
+	// If Annotations is set on the GatewayClass as well, the labels are merged with the Gateway's labels taking precedence.
+	//
+	// Support: Extended
+	Annotations map[string]string `json:"annotations,omitempty"`
+	// ParametersRef is a reference to a resource that contains the configuration
+	// parameters corresponding to the Gateway. This is optional if the
+	// controller does not require any additional configuration.
+	//
+	// ParametersRef can reference a standard Kubernetes resource, i.e. ConfigMap,
+	// or an implementation-specific custom resource. The resource must be namespace-scoped
+	// and live in the same namespace.
+	//
+	// If ParametersRef is set on both a Gateway and the corresponding GatewayClass, the merging
+	// behavior between the two are implementation-specific.
+	//
+	// Support: Implementation-specific
+	//
+	// +optional
+	ParametersRef *LocalParametersReference `json:"parametersRef,omitempty"`
+	...
+}
+
+type GatewayClassInfrastructure struct {
+	// Labels that should be applied to any resources created in response to this GatewayClass.
+	//
+	// For implementations creating other Kubernetes objects, this should be the `metadata.labels` field on resources.
+	// For other implementations, this refers to any relevant (implementation specific) "labels" concepts.
+	//
+	// If Labels is set on the Gateway as well, the labels are merged with the Gateway's labels taking precedence.
+	//
+	// Support: Extended
+	Labels map[string]string `json:"labels,omitempty"`
```

Comment:
Assuming we'll copy validation from upstream here, don't need to figure that out in this GEP, but it might be useful to at least describe the max number of entries in each map. I'd prefer to start with something small like 8 unless we have a clear use case for more.

Reply:
I put 16 max

Reply:
Any specific use cases that would benefit from more than 8?

Reply:
I just vaguely recall seeing some pretty monstrous set of annotations for AWS/Azure and it seemed pretty minimal risk to make it 16 tbh. So no strong case really.

Reply:
There definitely are some massive lists out there (for example https://kubernetes-sigs.github.io/aws-load-balancer-controller/v2.2/guide/service/annotations/), but I don't think it's common to configure all of these at the same time. I'd personally rather start with a smaller number here and only increase if/when needed. We've left room in the API to loosen validation in future releases, but not tighten it. 

I know it probably has limited efficacy, but somehow it feels like the tighter our validation is here, the less likely this field will end up being overloaded for unintended/unexpected purposes.